### DISCUSSION TITLE
Makie plotting

### DISCUSSION THREAD

**Discussion by koehlerson** [Ferrite-FEM/Ferrite.jl#313] [Ideas]
Hey,

I started to get into Makie plotting for our adaptive mesh student project and I think it would be nice to let user plot JuAFEM solutions in Makie. Often, when you start with computational problems you only need a rough feedback if things work as they should, i.e. no intense post processing. Nice side effect, you could also plot in notebooks with WGLMakie.

We could either provide a few recipes or directly dispatch `to_mesh` as in https://github.com/JuliaPlots/Plots.jl/issues/392#issuecomment-342238881 on `Grid` and `AbstractGrid`, respectively. It works quite well for me.

![2020-12-14_12-44](https://user-images.githubusercontent.com/27497290/102078134-b0b24400-3e0a-11eb-9af3-3664aa2af60d.png)

This should work for 3D problems as well (see http://juliaplots.org/MakieReferenceImages/gallery//fem_mesh_3d/index.html). 

Edit: I should mention that the plot windows are responsive to moving, zooming and in 3D rotating

Another Edit: My question to you guys is essentially: do you prefer a page at the docs, where we list some functions to convert stuff or do we directly incorporate plot recipes in the repo just as in: https://makie.juliaplots.org/dev/recipes.html. The latter would imply a Makie dependency.

**Comment by JimBrouzoulis** [comment]
I think this would be a very nice addition!

Adding some code snippets in the documentation would probably be the quickest solution and from my perspective will do for now. If we see that it starts being used by several people, then maybe it is better to add recipes. In this way, we don't need to make a decision, at the moment, if we want to add Makie as a dependency.

What about some of the other plotting libraries? I like plotly myself, can this be written so it works with something generic like Plots.jl?

**Comment by koehlerson** [comment]
So I have a branch where I included the dependency `AbstractPlotting`, there I dispatch functions like `convert_arguments` on `DofHandler` and a solution vector `u` and if its 1D you can call pretty much every point based plot function on it, i.e. `scatter`, `scatterlines`, `lines` and so on. Further, you have all the freedom to customize the plot which is very JuAFEM-ish(?)

Now in 2D and 3D it's a bit more complicated, there, I just overload plotting functions `AbstractPlotting.mesh` and `AbstractPlotting.surface`. The latter is only applicable for 2D cases and the former works for 2 and 3D. The benefit is that the dependency is quite small and user can choose their Makie backend (GLMakie if it needs to be fast, WGLMakie in notebooks). We only provide the blue print to convert the types such that any Makie backend can handle it. And still, you can customize everything as needed. Here are two gifs that show the repsonsive windows. Note that the original movements were smooth, the chopped off movements are a result of .gif framedrops
![mesh3d](https://user-images.githubusercontent.com/27497290/102232079-fc3e1e00-3eee-11eb-9686-63cbffd0d828.gif)
![surfaceplot](https://user-images.githubusercontent.com/27497290/102232090-ff390e80-3eee-11eb-8da4-5c54de1a6695.gif)
Unfortunately, I think that Plots.jl doesn't support `AbstractPlotting.jl` but I'm not sure on this.

**Reply by koehlerson** [reply]
https://github.com/KristofferC/JuAFEM.jl/pull/314 here it is

**Reply by JimBrouzoulis** [reply]
Looks really nice!
I won't have time to test this out the next few days, but maybe someone else will.

**Reply by koehlerson** [reply]
no worries, I need to take care of vector valued functions anyways. I'm trying to get a workaround for this, such that people get as default a magnitude plot, but can also pass a single component vector

**Reply by JimBrouzoulis** [reply]
Passing a function would be quite flexible, for example, a norm function or a function that picks out a given element from a vector

**Reply by koehlerson** [reply]
ok, should work now. Tested with the plasticity example you can do `mesh(dh, u; process=postprocess)` where `process` can be any function that takes an array of size `fielddims x nnodes` and maps it to an array of length `nnodes`. Default function `postprocess` takes the L2 norm of each column

**Reply by koehlerson** [reply]
the nice thing is that you can combine with all the other makie/abstractplotting functionality, e.g.

`vbox(hbox(mesh(dh,u; process=pick_x), mesh(dh,u; process=pick_y)), mesh(dh, u; process=pick_z))`
yields
![2020-12-16_15-28](https://user-images.githubusercontent.com/27497290/102361924-1130b500-3fb4-11eb-9fe3-a627c2edab84.png)

**Reply by koehlerson** [reply]
or even better: 

![2020-12-16_15-41](https://user-images.githubusercontent.com/27497290/102363568-d3cd2700-3fb5-11eb-8c63-a7efd67c5803.png)
